bookserver/routers/rslogging.py

```diff
@@ -179,3 +195,135 @@ async def runlog(request: Request, response: Response, data: LogRunIncoming):
 async def same_class(user1: AuthUserValidator, user2: str) -> bool:
     u2 = await fetch_user(user2)
     return user1.course_id == u2.course_id
+
+
+# completion tables
+# =================
+#
+# This section contains implementations of endpoints for tracking progress
+
+
+# updatelastpage
+# --------------
+# see :ref:`processPageState`
+@router.post("/updatelastpage")
+async def updatelastpage(request: Request, request_data: LastPageDataIncoming):
+    if request_data.last_page_url is None:
+        return  # todo:  log request_data, request.args and request.env.path_info
+    if request.state.user:
+        lpd = request_data.dict()
+        rslogger.debug(f"{lpd=}")
+
+        lpd["last_page_chapter"] = request_data.last_page_url.split("/")[-2]
```

Comment:
Document this nasty stuff. Ideally, hyperlink to the same hackery in the Runestone Components. Ick.

Reply:
Sometimes you just gotta parse a string man

Reply:
Added comments

Reply:
:) thanks!